### PR TITLE
JS: Add data-flow steps for arrays using a pseudo-property

### PR DESCRIPTION
**Pull request by erik-krogh**
I moved both the taint-steps and the new data-flow-steps for arrays into an `Arrays.qll` file. 

Gets us a TP for CVE-2018-3726. 

A [slightly outdated evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1583668336648) shows reasonable performance. 

The [new result from the evaluation](https://lgtm.com/query/6794730486992582712/) is a TP. 

I tried to see how our analysis worked if I removed the taint-steps and only relied on the new data-flow steps.  
The result was that we missed a lot of TPs due to taint-sources being arrays, and the data-flow steps need a write/read pair. 

[Here are examples of the new data-flow edges that are added](https://lgtm.com/query/3887672203853663035/).

### CODE REVIEW

**Comment by erik-krogh** [comment]
I'm fixing up the tests (also found a bug from that).  
And I'll look at using `DynamicPropertyAccess.qll` for the array reads.

File: javascript/ql/src/semmle/javascript/Arrays.qll [hunk 1]
```diff
@@ -92,3 +92,201 @@ module ArrayTaintTracking {
...
+   */
+  private class ArrayAppendStep extends DataFlow::AdditionalFlowStep, DataFlow::MethodCallNode {
+    ArrayAppendStep() {
+      this.getMethodName() = "push" or
+      this.getMethodName() = "unshift"
+    }
+
+    /**
+     * Holds if `pred` should be stored in the object `succ` under the property `prop`.
+     */
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/Arrays.qll]: `override` predicates inherit the qldoc and should rarely have a qldoc comment of their own.

File: javascript/ql/src/semmle/javascript/dataflow/Nodes.qll [hunk 2]
```diff
@@ -598,18 +598,23 @@ class ArrayConstructorInvokeNode extends DataFlow::InvokeNode {
  * new Array('apple', 'orange')
  * Array(16)
  * new Array(16)
+ * Array.from(1,2,3);
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/dataflow/Nodes.qll]: No it was not, that was a mistake from reading the taint steps on arrays a little too quickly.

**Comment by erik-krogh** [comment]
It seems like `EnumeratedPropName` from `DynamicPropertyAccess.qll` doesn't match how I would need to use it in this PR, unless `EnumeratedPropName` is refactored. 

In an ordinary `for(var i = 0; i < arr.length; i++){...}` loop there is no `EnumeratedPropName`, as the `i` is an index and not a property name.  
If we expand the `EnumeratedPropName` to include those kinds of loops, then it could work for this case, but otherwise I don't think it will. 

Also, I think I would need some more refactor, getting more focus on the enumerated object and the accesses on the object, rather than the property-name. 
Because there is not always an `EnumeratedPropName` to refer to. E.g. in the below example there is no `EnumeratedPropName`, but there is both a source-object and a source-property, which is what I'm interested in. 
```JavaScript
lodash.forEach(arr, (e) => sink(e))
```

**Comment by asgerf** [comment]
Sorry, I was thinking about the `getAnEnumeratedArrayElement` predicate in the same file, and whether it would make sense to share that code.

**Comment by erik-krogh** [comment]
> Sorry, I was thinking about the `getAnEnumeratedArrayElement` predicate in the same file, and whether it would make sense to share that code.

Yes it would, but I would have preferred to support a `EnumeratedPropName` refactorization, as it would give me support for libraries like `lodash.forEach`, `Object.keys(..)` and the others you got basically for free.

**Comment by erik-krogh** [comment]
> I'll use the getAnEnumeratedArrayElement

And that destroyed performance in `bwip-js`, I've reverted back to my previous approach where I search for an index declared as part of a `for` loop.

**Comment by erik-krogh** [comment]
A [very quick evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-esben.northeurope.cloudapp.azure.com_1583896675303) looks good.

**Comment by erik-krogh** [comment]
A more [detailed evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-esben.northeurope.cloudapp.azure.com_1584015119510) shows a less clear picture. 

Specifically the `sqlteaching` benchmark is bad.  
It looks to be the `js/insecure-randomness` query in particular that blows up. The `isAdditionalTaintStep` in that query blows up in combination with the new steps added by this PR. 
I'm looking into it.

**Comment by asgerf** [comment]
Since https://github.com/Semmle/ql/pull/3070 landed I guess the performance should be good now? Are you running a final evaluation?

**Comment by erik-krogh** [comment]
> Since #3070 landed I guess the performance should be good now? Are you running a final evaluation?

Yes, it should be ready tonight. 
But the preliminary results doesn't look all that positive.

**Comment by erik-krogh** [comment]
> Since #3070 landed I guess the performance should be good now? Are you running a final evaluation?

[This is the new evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1584560299525). 

I think the `sqlteaching` teaching benchmark can be ignored, the performance is way worse, but there is a good reason for that (the exploratory flow being about 4 orders of magnitude bigger). 

But even if we ignore `sqlteaching`, it doesn't look all that good.

**Comment by asgerf** [comment]
I have a branch that makes exploratory flow more precise. I've confirmed that merging it with your PR eliminates the overhead you observed in `sqlteaching` (it becomes the same as current master).

As for the rest of the evaluation, there's obviously a biased noise in the wall clock timings, but it could be hiding a genuine slowdown. Could you pick one of the slowest ones from that report other than `sqlteaching` and test it in isolation to see the actual overhead?

**Comment by erik-krogh** [comment]
> As for the rest of the evaluation, there's obviously a biased noise in the wall clock timings, but it could be hiding a genuine slowdown. Could you pick one of the slowest ones from that report other than `sqlteaching` and test it in isolation to see the actual overhead?

Here is an evaluation of just one of the slowest ones: https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-esben.northeurope.cloudapp.azure.com_1584809774080

**Pull request merged**